src/factor_graph.cc

```diff
@@ -159,11 +156,15 @@ weight_id_t CompactFactorGraph::get_categorical_weight_id(
    * The weight index is
    * (...((((0 * d1 + i1) * d2) + i2) * d3 + i3) * d4 + ...) * dk + ik
    *
-   * For FUNC_AND_CATEGORICAL, we look up the weight_ids map.
+   * For FUNC_AND_CATEGORICAL, we look up the factor_params map.
    *
    * TODO: refactor the above formula into a shared routine. (See also
    * binary_format.cc read_factors)
    */
+
+  static constexpr FactorParams INVALID_PARAMS = {Weight::INVALID_ID,
+                                                  DEFAULT_FEATURE_VALUE};
```

Comment:
Maybe this should go to factor.h? next to or inside the type as `FactorParams::INVALID_PARAMS`